### DISCUSSION TITLE
Erro ao Incluir [Authorize] AccountController

### DISCUSSION THREAD

**Discussion by wmnogueira** [balta-io/2811#31] [Q&A] [answered]
Ola Balta, tudo bem por ai?? Bom, acompanhando seu curso ([Fundamentos do ASP.NET 6 (100%) e tentando trazer para minha necessidade, encontrei um contra tempo em adicionar a Regra para autorizar o Role em Usuario. Ao aplicar o [Authorize] o Get passa a dar o erro:

![image](https://user-images.githubusercontent.com/66086669/184502288-3c46dd3e-9b5d-4e66-80b3-3c701d3687ea.png)

Segue abaixo a seguencia do codigo

Classe AccountController
`using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using WebPix.Services;

namespace WebPix.Controllers
{
    /// <summary>
    /// Introdução: https://balta.io/player/assistir/8407be55-4137-48ae-baf5-470eef47a297/38598d9f-1adb-4638-8a4b-9c77ca3b9bc4
    /// </summary>
    
    [ApiController]
    public class AccountController:ControllerBase
    {
        /// <summary>
        /// Estudar sobre Controle de Dependencia
        /// </summary>
        [AllowAnonymous]
        [HttpPost("v1/login")]
        public IActionResult Login(
            [FromServices]TokenService tokenService
            )
        {
            var token = tokenService.GenerateToken(null);

            return Ok(token);
        }

        
        [HttpGet("v1/usuario")]
        [Authorize(Roles = "admin")]
        public IActionResult GetUsuario() => Ok(User.Identity.Name);

        [HttpGet("v1/autor")]
        public IActionResult GetAutor() => Ok(User.Identity.Name);

        [HttpGet("v1/admin")]
        public IActionResult GetAdmin() => Ok(User.Identity.Name);
    }
}`

`using System.Security.Claims;
using System.Text;
using WebPix.Models.Sistema;

namespace WebPix.Services
{
    public class TokenService
    {
        /// <summary>
        /// Introdução: https://balta.io/player/assistir/8407be55-4137-48ae-baf5-470eef47a297/85412406-29b3-4592-a631-e76c2162dad8
        /// </summary>
        /// <param name="user"></param>
        /// <returns></returns>
        public string GenerateToken(Usuario user)
        {
            var tokenHandler = new JwtSecurityTokenHandler();
            var key = Encoding.ASCII.GetBytes(Configuration.JwtKey);

            //TokenDescriptor: https://balta.io/player/assistir/8407be55-4137-48ae-baf5-470eef47a297/85412406-29b3-4592-a631-e76c2162dad8
            var tokenDescriptor = new SecurityTokenDescriptor
            {
                //Definição de Clain: AFIRMAÇÃO https://balta.io/player/assistir/8407be55-4137-48ae-baf5-470eef47a297/a5b721c0-7317-45a7-8558-a953b4f4fc43
                Subject = new ClaimsIdentity(new Claim[]
                {
                    new Claim(ClaimTypes.Name, "marcelinho"),
                    new Claim(ClaimTypes.Role, "admin")
                }),

                Expires = DateTime.UtcNow.AddHours(8),
                SigningCredentials = new SigningCredentials(
                    new SymmetricSecurityKey(key),
                    algorithm:SecurityAlgorithms.HmacSha256Signature),

            };

            var token = tokenHandler.CreateToken(tokenDescriptor);

            return tokenHandler.WriteToken(token);
        }

    }
}`

`using System.Text;
using WebPix;
using System.Text.Json.Serialization;
using WebPix.Data;
using WebPix.Services;
using Microsoft.AspNetCore.Authentication.JwtBearer;
using Microsoft.IdentityModel.Tokens;

var builder = WebApplication.CreateBuilder(args);

/// <summary>
/// Introdução: https://balta.io/player/assistir/8407be55-4137-48ae-baf5-470eef47a297/36362ee4-42ae-4950-837a-b9f9ffa249e8
/// </summary>
var key = Encoding.ASCII.GetBytes(Configuration.JwtKey);

builder
    .Services.AddAuthentication(x =>
    {
        x.DefaultAuthenticateScheme = JwtBearerDefaults.AuthenticationScheme;
        x.DefaultChallengeScheme = JwtBearerDefaults.AuthenticationScheme;
    }).AddJwtBearer(x =>
    {
        x.TokenValidationParameters = new TokenValidationParameters
        {
            ValidateIssuerSigningKey = true,
            IssuerSigningKey = new SymmetricSecurityKey(key),
            ValidateIssuer = false,
            ValidateAudience = false
        };
    });

builder
    .Services
    .AddControllers()
    .ConfigureApiBehaviorOptions(options =>
    {
        options.SuppressModelStateInvalidFilter = true;
    });

builder
    .Services
    .AddControllers()
    //Evitar validação automatica ModelState: https://balta.io/player/assistir/8407be55-4137-48ae-baf5-470eef47a297/af71e102-00d1-49b1-84b1-79a40e08ea95
    .ConfigureApiBehaviorOptions(options =>
    {
        options.SuppressModelStateInvalidFilter = true;
    })
     .AddJsonOptions(x =>
     {
         x.JsonSerializerOptions.ReferenceHandler = ReferenceHandler.IgnoreCycles;
         x.JsonSerializerOptions.DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingDefault;
     });

builder
    .Services
    .AddDbContext<PixDataContext>();

///Explicação do Tempo de Uso do Token
///https://balta.io/player/assistir/8407be55-4137-48ae-baf5-470eef47a297/64e00da2-745f-4cc6-b213-9fbd10ebf76b
builder
    .Services
    .AddTransient<TokenService>();
    

var app = builder.Build();

app.MapControllers();

app.Run();`

Poderia me dar uma "Luz" donde estou errando??

Marcelinho

**Comment by andrebaltieri** [comment, marked as answer]
@wmnogueira pelo erro, ele diz que o tipo de autenticação que você pediu e usou não são compatíveis... por exemplo, você disse que usaria autenticação via Cookies e partiu para o JWT Bearer.... só um exemplo...

Minha dica:
Cria um projeto separado e faz só esta parte da autenticação... Começa do básico, só colocando os roles, hard coded mesmo e vai evoluindo até chegar no que precisa...

Normalmente eu faço isto e funciona... além do debug é claro... mas neste caso o DeBug não vai te falar algo muito útil acredito eu...

**Reply by wmnogueira** [reply]
Boa tardeee!! Balta seguindo sua dica em criar um projeto a Parte acabei conseguindo identificar aonde estava o problema. O que lhes fala acabou deixando de incluir em program as linhas de:

app.Authentication()
app.Authorization()

Mas valeu mais uma vez como aprendizado.

Grato pela atenção e dica.

Marcelinho

**Comment by wmnogueira** [comment]
ok, Obrigado pela dica
________________________________
De: Andre Baltieri ***@***.***>
Enviado: segunda-feira, 15 de agosto de 2022 10:05
Para: balta-io/2811 ***@***.***>
Cc: Marcelo Nogueira ***@***.***>; Mention ***@***.***>
Assunto: Re: [balta-io/2811] Erro ao Incluir [Authorize] AccountController (Discussion #31)


@wmnogueira<https://github.com/wmnogueira> pelo erro, ele diz que o tipo de autenticação que você pediu e usou não são compatíveis... por exemplo, você disse que usaria autenticação via Cookies e partiu para o JWT Bearer.... só um exemplo...

Minha dica:
Cria um projeto separado e faz só esta parte da autenticação... Começa do básico, só colocando os roles, hard coded mesmo e vai evoluindo até chegar no que precisa...

Normalmente eu faço isto e funciona... além do debug é claro... mas neste caso o DeBug não vai te falar algo muito útil acredito eu...
